tap_googleads/custom_query_stream.py

```diff
@@ -0,0 +1,178 @@
+DATE_TYPES = ("segments.date", "segments.month", "segments.quarter", "segments.week")
+
+
+class CustomQueryStream(ReportsStream):
+    """Define custom stream."""
+
+    records_jsonpath = "$.results[*]"
+    primary_keys = []
+    replication_key = None
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Initialize the REST stream.
+
+        Args:
+            tap: Singer Tap this stream belongs to.
+        """
+        self.custom_query = kwargs.pop("custom_query")
+        self._query = self.custom_query["query"]
+        self.name = self.custom_query["name"]
+        super().__init__(*args, **kwargs)
+
+    @property
+    def gaql(self):
+        return self._query
+
+    @cached_property
+    def schema(self) -> dict:
+        """Return dictionary of record schema.
+
+        Dynamically detect the json schema for the stream.
+        This is evaluated prior to any records being retrieved.
+        """
+        local_json_schema = {
+            "type": "object",
+            "properties": {},
+            "additionalProperties": True,
+        }
+
+        google_datatype_mapping = {
+            "STRING": "string",
+            "MESSAGE": "string",
+            "BOOLEAN": "boolean",
+            "DATE": "string",
+            "ENUM": "string",
+            "INT64": "integer",
+            "INT32": "integer",
+            "DOUBLE": "number",
+        }
+        try:
+            query_object = sqlparse.parse(self.custom_query["query"])[0]
+        except ValueError:
+            message = f"The custom GAQL query {self.custom_query['name']} failed. Validate your GAQL query with the Google Ads query validator. https://developers.google.com/google-ads/api/fields/v19/query_validator"
+            raise Exception(message)
+
+        fields = []
+        has_where_clause = False
+        for token in query_object.tokens:
+            if isinstance(token, sqlparse.sql.IdentifierList):
+                fields = [field.strip() for field in token.value.split(",")]
+            if isinstance(token, sqlparse.sql.Where):
+                has_where_clause = True
+
+        date_filter = (
+            f"segments.date >= {self.start_date} and segments.date <= {self.end_date}"
+        )
+        if "segments.date" not in fields:
+            fields.append("segments.date")
+        if has_where_clause:
+            self._query = self._query + f" AND {date_filter}"
+        else:
+            self._query = self._query + f" WHERE {date_filter}"
+
+        google_schema = self.get_fields_metadata(fields)
+
+        for field in fields:
+            node = google_schema[field]
+            google_data_type = node.get("dataType", "")
+            field_value = {
+                "type": [
+                    google_datatype_mapping.get(google_data_type, "string"),
+                    "null",
+                ]
+            }
+
+            if google_data_type == "DATE" and field in DATE_TYPES:
+                field_value["format"] = "date"
+
```

Comment:
Is the check against `DATE_TYPES` necessary? Surely if the data type is coming through as `DATE` we should treat it as a date regardless?

```suggestion
```